src/platform/datapath_raw_socket.c

```diff
@@ -204,47 +206,188 @@ CxPlatTryAddSocket(
             MappedAddress.Ipv6.sin6_family = AF_INET6;
         }
 #endif
+        if (Socket->UseTcp) {
+            //
+            // Create a temporary UDP socket bound to a wildcard port
+            // and connect this socket to the remote address.
+            // By doing this, the OS will select a local address for us.
```

Comment:
Whatever port the stack assigned for UDP isn't guaranteed to be available for TCP though, right?

Reply:
That doesn't matter. The port chosen doesn't affect source routing. The TCP port has already been chosen before the code reaches here and that's the port will be eventually using. This temporary UDP port is only for `connect`.